### PR TITLE
Add Tests for the Relaxed Data Binding

### PR DESCRIPTION
**Pull request by SachinAkash01**
Part of: https://github.com/ballerina-platform/ballerina-library/issues/7366

## Examples

N/A

## Checklist
- [x] Linked to an issue
- [ ] ~Updated the changelog~
- [x] Added tests
- [ ] ~Updated the spec~
- [x] Checked native-image compatibility
- [ ] ~Checked the impact on OpenAPI generation~

### CODE REVIEW

File: ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_tests.bal
Comment (TharmiganK) [inline ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_tests.bal:1]: Shall we have a test case for projection? For example, use a **closed** record for service/client side binding but sent an actual json with **fields other than specified in the type**.

File: ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_tests.bal [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+|};
+
+type Teacher record {
+    int id;
+    string name;
+    string? subject;
+};
+
+listener http:Listener laxDataBindingEP = new (laxDataBindingTestPort, httpVersion = http:HTTP_2_0);
+final http:Client laxClient = check new ("http://localhost:" + laxDataBindingTestPort.toString(), {
```
Comment (TharmiganK) [inline ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_tests.bal:39]: Lets have negative test cases with normal HTTP client where all these special relaxed scenarios should fail

File: ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_tests.bal [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        "rank": "1",
+        "classification": "unknown"
+    });
+    test:assertEquals(reponse, expectedPayload);
+}
+
+@test:Config
+function testMissingRequiredField() returns error? {
+    Teacher expectedPayload = {"id": 3001, "name": "Anna", "subject": null};
+    Teacher response = check laxClient->/people/test3.post({ "id": 3001, "name": "Anna"});
```
Comment (TharmiganK) [inline ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_tests.bal:107]: Formatting issue:
```suggestion
    Teacher response = check laxClient->/people/test3.post({"id": 3001, "name": "Anna"});
```

Check other places as well

File: ballerina-tests/http-advanced-tests/tests/test_service_ports.bal [hunk 3]
```diff
@@ -50,3 +50,5 @@ const int http1SsePort = 9094;
 const int http2SsePort = 9095;
 
 const int serviceContractTestPort = 9096;
+
+const int laxDataBindingTestPort = 9015;
```
Comment (TharmiganK) [inline ballerina-tests/http-advanced-tests/tests/test_service_ports.bal:54]: Lets change the port
```suggestion
const int laxDataBindingTestPort = 9097;
```

File: native/src/main/java/io/ballerina/stdlib/http/api/service/signature/converter/JsonToRecordConverter.java [hunk 4]
```diff
@@ -76,15 +76,16 @@ private static Object getRecord(Type entityBodyType, Object bJson, boolean laxDa
             Map<String, Object> valueMap = new HashMap<>();
             Boolean bool = Boolean.FALSE;
             valueMap.put(ENABLE_CONSTRAINT_VALIDATION, bool);
```
Comment (TharmiganK) [inline native/src/main/java/io/ballerina/stdlib/http/api/service/signature/converter/JsonToRecordConverter.java:78]:
```suggestion
            valueMap.put(ENABLE_CONSTRAINT_VALIDATION, Boolean.FALSE);
```

File: native/src/main/java/io/ballerina/stdlib/http/api/service/signature/converter/JsonToRecordConverter.java [hunk 5]
```diff
@@ -76,15 +76,16 @@ private static Object getRecord(Type entityBodyType, Object bJson, boolean laxDa
...
-            }
             BMap<BString, Object> mapValue = ValueCreator.createRecordValue(
                     io.ballerina.lib.data.ModuleUtils.getModule(),
                     PARSER_AS_TYPE_OPTIONS, valueMap);
+            if (laxDataBinding) {
+                BMap allowDataProjection = mapValue.getMapValue(ALLOW_DATA_PROJECTION);
+                allowDataProjection.put(NIL_AS_OPTIONAL, Boolean.TRUE);
+                allowDataProjection.put(ABSENT_AS_NILABLE, Boolean.TRUE);
+            } else {
+                mapValue.put(ALLOW_DATA_PROJECTION, bool);
```
Comment (TharmiganK) [inline native/src/main/java/io/ballerina/stdlib/http/api/service/signature/converter/JsonToRecordConverter.java:87]:
```suggestion
                mapValue.put(ALLOW_DATA_PROJECTION, Boolean.FALSE);
```

File: native/src/main/java/io/ballerina/stdlib/http/api/service/signature/converter/JsonToRecordConverter.java [hunk 6]
```diff
@@ -76,15 +76,16 @@ private static Object getRecord(Type entityBodyType, Object bJson, boolean laxDa
             Map<String, Object> valueMap = new HashMap<>();
             Boolean bool = Boolean.FALSE;
```
Comment (TharmiganK) [inline native/src/main/java/io/ballerina/stdlib/http/api/service/signature/converter/JsonToRecordConverter.java:77]: Lets remove this local variable

File: ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_tests.bal
Comment (SachinAkash01) [reply ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_tests.bal:1]: I think what you have mentioned is this right? https://github.com/ballerina-platform/module-ballerina-http/pull/2259/files#diff-30d871a0e9728f384bfbd24680a15d6b671ef4c50d99c20a022a0510dd43bbf1R91-R102

Comment (TharmiganK) [reply ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_tests.bal:1]: Yes, Do we have test in the client side as well?

Comment (SachinAkash01) [reply ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_tests.bal:1]: > Yes, Do we have test in the client side as well?

No I will add a test for the client side as well

Comment (SachinAkash01) [reply ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_tests.bal:1]: Add a test for projection in client side binding with this commit: [891b9e200840639302056c3d6699d3f3ddb40dc9](https://github.com/ballerina-platform/module-ballerina-http/pull/2259/commits/891b9e200840639302056c3d6699d3f3ddb40dc9)

File: ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_tests.bal (hunk 1 above)
Comment (SachinAkash01) [reply ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_tests.bal:39]: Added negative tests with this commit: [f2c9eab9854ad8dc310c730d1a2909e42f1780d1](https://github.com/ballerina-platform/module-ballerina-http/pull/2259/commits/f2c9eab9854ad8dc310c730d1a2909e42f1780d1)

File: ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_negative_tests.bal [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            "address": "Piliyandala",
+            "Experience": 10
+        };
+    }
+}
+
+@test:Config
+function testNullForOptionalFieldNegative() returns error? {
+    Employee|error response = laxNegativeClient->/company/test1.post({ "id": 1001, "name": "Tharmigan", "address": null});
+    if (response is http:ClientRequestError) {
```
Comment (TharmiganK) [inline ballerina-tests/http-advanced-tests/tests/relaxed_data_binding_negative_tests.bal:83]: No need for parentheses here:
```suggestion
    if response is http:ClientRequestError {
```

Lets try to follow the best practices: https://learn-ballerina.github.io/best_practices/language_best_practices.html

Check other places as well

**Pull request merged**